### PR TITLE
Add tests for export

### PR DESCRIPTION
**Pull request by ushitora-anqou**
~~The commits included in this PR should be rebased after #54 is merged into the main branch.~~ done.

### CODE REVIEW

File: internal/controller/mantlebackup_controller_test.go [hunk 1]
```diff
@@ -489,6 +489,13 @@ var _ = Describe("MantleBackup controller", func() {
 			Expect(*jobExport.Spec.Template.Spec.SecurityContext.RunAsGroup).To(Equal(int64(10000)))
 			Expect(*jobExport.Spec.Template.Spec.SecurityContext.RunAsNonRoot).To(Equal(true))
 
+			// Make sure FROM_SNAP_NAME is empty because we're performing a full backup.
+			for _, evar := range jobExport.Spec.Template.Spec.Containers[0].Env {
+				if evar.Name == "FROM_SNAP_NAME" {
+					Expect(evar.Value).To(Equal(""))
+				}
```
Comment (satoru-takeuchi) [inline internal/controller/mantlebackup_controller_test.go:501]: Could you also check the `FROM_SNAP_NAME` is defined in `Spec.Containers[0].Env`? The current code can't distinguish the following cases:

- (a) This envvar is not defined.
- (b) This envvar is defined and is empty.

File: internal/controller/mantlebackup_controller_test.go [hunk 2]
```diff
@@ -541,6 +548,25 @@ var _ = Describe("MantleBackup controller", func() {
...
+				},
+				&jobExport2,
+			)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Make sure FROM_SNAP_NAME is filled correctly because we're performing an incremental backup.
+			for _, evar := range jobExport2.Spec.Template.Spec.Containers[0].Env {
+				if evar.Name == "FROM_SNAP_NAME" {
+					Expect(evar.Value).To(Equal(backup.GetName()))
+				}
```
Comment (satoru-takeuchi) [inline internal/controller/mantlebackup_controller_test.go:584]: ditto.

File: internal/controller/mantlebackup_controller_test.go [hunk 3]
```diff
@@ -496,6 +497,18 @@ var _ = Describe("MantleBackup controller", func() {
...
+			var jobUpload batchv1.Job
+			err = k8sClient.Get(
+				ctx,
+				types.NamespacedName{
+					Name:      fmt.Sprintf("mantle-upload-%s", backup.GetUID()),
+					Namespace: resMgr.ClusterID,
+				},
+				&jobUpload,
+			)
+			Expect(aerrors.IsNotFound(err)).To(BeTrue())
```
Comment (satoru-takeuchi) [inline internal/controller/mantlebackup_controller_test.go:514]: Please use `Consistently` here not to overlook the creation of upload job after #510.

File: internal/controller/mantlebackup_controller_test.go [hunk 4]
```diff
@@ -894,3 +937,431 @@ var _ = Describe("prepareForDataSynchronization", func() {
 		),
 	)
 })
+
+var _ = Describe("SetSynchronizing", func() {
+	doTestCallOnce := func(
+		target, source *mantlev1.MantleBackup,
+		isError bool,
```
Comment (satoru-takeuchi) [inline internal/controller/mantlebackup_controller_test.go:944]:
```suggestion
		shouldBeError bool,
```

File: internal/controller/mantlebackup_controller_test.go [hunk 5]
```diff
@@ -894,3 +937,431 @@ var _ = Describe("prepareForDataSynchronization", func() {
...
+				}
+				diffTo, ok := source.GetAnnotations()[annotDiffTo]
+				if !ok || diffTo != target.GetName() {
+					return errors.New("diffTo is invalid")
+				}
+				return nil
+			},
+		),
+		Entry(
+			"a backup should fail if target's ReadToUse is True",
```
Comment (satoru-takeuchi) [inline internal/controller/mantlebackup_controller_test.go:1054]:
```suggestion
			"a backup should fail if target's ReadyToUse is True",
```

File: internal/controller/mantlebackup_controller_test.go [hunk 6]
```diff
@@ -894,3 +937,431 @@ var _ = Describe("prepareForDataSynchronization", func() {
...
+			func(_, _ *mantlev1.MantleBackup) error { return nil },
+		),
+	)
+
+	doTestCallTwice := func(
+		name1 string,
+		diffFrom1 *string,
+		name2 string,
+		diffFrom2 *string,
+		isError bool,
```
Comment (satoru-takeuchi) [inline internal/controller/mantlebackup_controller_test.go:1076]:
```suggestion
		shouldBeError bool,
```

**Review by satoru-takeuchi — CHANGES_REQUESTED** [review]
I completed to review this PR and left some additional minor suggestions.

File: internal/controller/mantlebackup_controller_test.go [hunk 7]
```diff
@@ -894,3 +937,431 @@ var _ = Describe("prepareForDataSynchronization", func() {
...
+		// make sure that only 1 Job is created
+		Consistently(ctx, func(g Gomega) error {
+			numJobs, err := getNumOfExportJobs(nsController)
+			g.Expect(err).NotTo(HaveOccurred())
+			g.Expect(numJobs).To(Equal(1))
+			return nil
+		}, "1s").Should(Succeed())
+
+		// make sure that another mantle-controller existing in a different namespace can create an export Job.
+		ns2 := resMgr.CreateNamespace()
```
[the remark this answers is not in the document]
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller_test.go]: Fixed in https://github.com/cybozu-go/mantle/pull/59/commits/7078d9890755fe72881e7618af2cbfec012b14f3.

File: internal/controller/mantlebackup_controller_test.go [hunk 8]
```diff
@@ -894,3 +937,431 @@ var _ = Describe("prepareForDataSynchronization", func() {
...
+			&PrimarySettings{
+				Client:                 grpcClient,
+				ExportDataStorageClass: resMgr.StorageClassName,
+				MaxExportJobs:          1,
+			},
+			"dummy image",
+			"",
+			nil,
+		)
+		createAndExportMantleBackup(mbr2, "target2", ns2)
```
[the remark this answers is not in the document]
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller_test.go:1369]: Fixed in https://github.com/cybozu-go/mantle/pull/59/commits/7078d9890755fe72881e7618af2cbfec012b14f3.

File: internal/controller/mantlebackup_controller_test.go (hunk 1 above)
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller_test.go:503]: Fixed in https://github.com/cybozu-go/mantle/pull/59/commits/aefbb65b37389ee148bac132588d9573da18af4a.

File: internal/controller/mantlebackup_controller_test.go (hunk 2 above)
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller_test.go:592]: Fixed in https://github.com/cybozu-go/mantle/pull/59/commits/aefbb65b37389ee148bac132588d9573da18af4a.

File: internal/controller/mantlebackup_controller_test.go (hunk 3 above)
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller_test.go]: Fixed in https://github.com/cybozu-go/mantle/pull/59/commits/c07278f10481efe9bcef0cca114a7b090f7b1380.

File: internal/controller/mantlebackup_controller_test.go (hunk 4 above)
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller_test.go]: Fixed in https://github.com/cybozu-go/mantle/pull/59/commits/2cde0591eb1b568fd990cc709bd7619822853972.

File: internal/controller/mantlebackup_controller_test.go (hunk 5 above)
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller_test.go]: Fixed in https://github.com/cybozu-go/mantle/pull/59/commits/2cde0591eb1b568fd990cc709bd7619822853972.

File: internal/controller/mantlebackup_controller_test.go (hunk 6 above)
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller_test.go]: Fixed in https://github.com/cybozu-go/mantle/pull/59/commits/2cde0591eb1b568fd990cc709bd7619822853972.

**Comment by ushitora-anqou** [comment]
@satoru-takeuchi I fixed you points. Could you review this again?

**Review by satoru-takeuchi — APPROVED** [review]
LGTM. Please squash commits.

**Comment by ushitora-anqou** [comment]
@satoru-takeuchi I squashed my commits. Could you review this PR again?

**Pull request merged**